README.md
Neste projeto irei implementar o jogo "Sudoku" utilizando a linguagem de programação Python.

O Sudoku consiste em um tabuleiro 9x9 onde:
-Cada linha deve possuir apenas uma vez os número de 0 a 9.
-Cada coluna deve possuir apenas uma vez os número de 0 a 9.
-Cada bloco 3x3 deve possuir apenas uma vez os número de 0 a 9.


history.txt
#Rows
        ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9'],
        ['B1', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B9'],
        ['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9'],
        ['D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9'],
        ['E1', 'E2', 'E3', 'E4', 'E5', 'E6', 'E7', 'E8', 'E9'],
        ['F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9'],
        ['G1', 'G2', 'G3', 'G4', 'G5', 'G6', 'G7', 'G8', 'G9'],
        ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'H7', 'H8', 'H9'],
        ['I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9'],

        #Cols
        ['A1', 'B1', 'C1', 'D1', 'E1', 'F1', 'G1', 'H1', 'I1'],
        ['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2'],
        ['A3', 'B3', 'C3', 'D3', 'E3', 'F3', 'G3', 'H3', 'I3'],
        ['A4', 'B4', 'C4', 'D4', 'E4', 'F4', 'G4', 'H4', 'I4'],
        ['A5', 'B5', 'C5', 'D5', 'E5', 'F5', 'G5', 'H5', 'I5'],
        ['A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6', 'I6'],
        ['A7', 'B7', 'C7', 'D7', 'E7', 'F7', 'G7', 'H7', 'I7'],
        ['A8', 'B8', 'C8', 'D8', 'E8', 'F8', 'G8', 'H8', 'I8'],
        ['A9', 'B9', 'C9', 'D9', 'E9', 'F9', 'G9', 'H9', 'I9'],

        #Blocks
        ['A1', 'A2', 'A3', 'B1', 'B2', 'B3', 'C1', 'C2', 'C3'],
        ['A4', 'A5', 'A6', 'B4', 'B5', 'B6', 'C4', 'C5', 'C6'],
        ['A7', 'A8', 'A9', 'B7', 'B8', 'B9', 'C7', 'C8', 'C9'],
        ['D1', 'D2', 'D3', 'E1', 'E2', 'E3', 'F1', 'F2', 'F3'],
        ['D4', 'D5', 'D6', 'E4', 'E5', 'E6', 'F4', 'F5', 'F6'],
        ['D7', 'D8', 'D9', 'E7', 'E8', 'E9', 'F7', 'F8', 'F9'],
        ['G1', 'G2', 'G3', 'H1', 'H2', 'H3', 'I1', 'I2', 'I3'],
        ['G4', 'G5', 'G6', 'H4', 'H5', 'H6', 'I4', 'I5', 'I6'],
        ['G7', 'G8', 'G9', 'H7', 'H8', 'H9', 'I7', 'I8', 'I9'],

count_test.py
#!/usr/bin/env python
# Four spaces as indentation [no tabs]

class Count_Test:

    def __init__(self):
        pass

    def cross(self, A, B):

        return [a+b for a in A for b in B]


    def printable(self):

        cols = '123456789'
        rows = 'ABCDEFGHI'
        squares = self.cross(rows, cols)
        unitlist = ([self.cross(rows, c) for c in cols] +
                    [self.cross(r, cols) for r in rows] +
                    [self.cross(rs, cs) for rs in ('ABC','DEF','GHI') for cs in ('123','456','789')])
        units = dict((s, [u for u in unitlist if s in u]) for s in squares)
        peers = dict((s, set(sum(units[s],[]))-set([s])) for s in squares)

        print units, "UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU"
        print peers, "PPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP"
# ==========================================
# Main
# ==========================================

if __name__ == "__main__":
    test = Count_Test()
    test.printable()

util.py
#!/usr/bin/env python
# Four spaces as indentation [no tabs]

# ------------------------------------------
# Find collision
# ------------------------------------------

#numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]

def find_collision(board, peers, cell, number):

    # -Dictionary with peers:
    # List of every cell that is assigned in a unit with this cell, except itself.
    # Example: I6 = [A6, B6, C6, D6, E6, F6, G6, H6, I1, I2, I3, I4, I5, I7, I8, I9, G4, G5, H4, H5]
    # Code: self.peers = collections.OrderedDict((s, set(sum(self.units[s],[]))-set([s])) for s in self.squares)

    #Check if a dict of peers have collision of numbers
    #for n in numbers:
    for p in list(peers[cell]):
        if board[p] == number:
            return True

    return False


# ------------------------------------------
# Find empty cells
# ------------------------------------------

def find_empty_cells(board, squares):

    return [ept_cl for ept_cl in squares if board[ept_cl] is 0]


# ------------------------------------------
# Find empty neighbors
# ------------------------------------------

def find_empty_peers(board, peers, cell):

    empty_peers = []

    #return [p for p in peers if peers[cell] is 0]
    for p in list(peers[cell]):
        if board[p] == 0:
            empty_peers.append(board[p])

    return empty_peers

# ------------------------------------------
# Format the print_board function output
# ------------------------------------------

def __format__(self, format):

    if (format == '0'):
        return '.'


# ------------------------------------------
# Print board
# ------------------------------------------

def print_board(board):

    # .format() -> format string using the board cells
    # .replace() -> format string changing '0' into '.' 

    print """
    1 2 3   4 5 6   7 8 9
    ---------------------
 A| {b[A1]} {b[A2]} {b[A3]} | {b[A4]} {b[A5]} {b[A6]} | {b[A7]} {b[A8]} {b[A9]} |
 B| {b[B1]} {b[B2]} {b[B3]} | {b[B4]} {b[B5]} {b[B6]} | {b[B7]} {b[B8]} {b[B9]} |
 C| {b[C1]} {b[C2]} {b[C3]} | {b[C4]} {b[C5]} {b[C6]} | {b[C7]} {b[C8]} {b[C9]} |
  | ------+-------+------ |
 D| {b[D1]} {b[D2]} {b[D3]} | {b[D4]} {b[D5]} {b[D6]} | {b[D7]} {b[D8]} {b[D9]} |
 E| {b[E1]} {b[E2]} {b[E3]} | {b[E4]} {b[E5]} {b[E6]} | {b[E7]} {b[E8]} {b[E9]} |
 F| {b[F1]} {b[F2]} {b[F3]} | {b[F4]} {b[F5]} {b[F6]} | {b[F7]} {b[F8]} {b[F9]} |
  | ------+-------+------ |
 G| {b[G1]} {b[G2]} {b[G3]} | {b[G4]} {b[G5]} {b[G6]} | {b[G7]} {b[G8]} {b[G9]} |
 H| {b[H1]} {b[H2]} {b[H3]} | {b[H4]} {b[H5]} {b[H6]} | {b[H7]} {b[H8]} {b[H9]} |
 I| {b[I1]} {b[I2]} {b[I3]} | {b[I4]} {b[I5]} {b[I6]} | {b[I7]} {b[I8]} {b[I9]} |
    ---------------------
""".format(b=board).replace('0', '.')

common.py
#!/usr/bin/env python
# Four spaces as indentation [no tabs]

import os, sys, inspect
import threading

PATH          = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))

# Interface
WAIT          = 1
ZOOM          = 6
TILE_WIDTH    = 16 * ZOOM
TILE_HEIGHT   = 16 * ZOOM

# ------------------------------------------
# Movements Exceptions
# ------------------------------------------

class BlockedMovementException(Exception):
    def __init__(self, board, cell):
        self.board = list(board)
        self.cell = cell
    def __str__(self):
        return "Blocked cell: {}.".format(self.cell)

class InvalidMovementException(Exception):
    def __init__(self, board, cell):
        self.board = list(board)
        self.cell = cell
    def __str__(self):
        return "Invalid cell: {}.".format(self.cell)

class NoMovementException(Exception):
    def __init__(self, board):
        self.board = list(board)
    def __str__(self):
        return "Board has no other possible movement."

class TimeoutError(Exception): pass

def timelimit(timeout):
    def internal(function):
        def internal2(*args, **kw):
            class Calculator(threading.Thread):
                def __init__(self):
                    threading.Thread.__init__(self)
                    self.result = None
                    self.error = None

                def run(self):
                    try:
                        self.result = function(*args, **kw)
                    except:
                        self.error = sys.exc_info()[0]

            c = Calculator()
            c.start()
            if timeout > 0:
                c.join(timeout)
            else:
                c.join()
            if c.isAlive():
                raise TimeoutError
            if c.error:
                raise c.error
            return c.result
        return internal2
    return internal

ai_player.py
#!/usr/bin/env python
# Four spaces as indentation [no tabs]

from util import *


# ==========================================
# Artificial Intelligence Player
# ==========================================

true_empty_cells = []

class AI_Player():

    # ------------------------------------------
    # Initialize
    # ------------------------------------------

    def __init__(self, board, unitlist, peers, squares):
        self.board = board
        self.unitlist = unitlist
        self.peers = peers
        self.squares = squares


    # ------------------------------------------
    # Check movement
    # ------------------------------------------

    def check_solving(self):

        collided = find_collision(self.board) != ([], 0)
        print "COLIDIU?", collided, "FIND_COLLISION:", find_collision(self.board)
        if collided:
            return (-1)
        else:
            print "WIBIDIBOBIDI DU"
            return (1)


    # ------------------------------------------
    # Try to solve the Sudoku
    # Insert 1 in the cell
    # ------------------------------------------

    def solve(self, ept_cells):

        number = 1

        while not bool(ept_cells):

            next_move = ept_cells.pop(0)

            self.board[next_move] = number

            

            self.next_number(number)

        return 1


    # ------------------------------------------
    # Generate leafs of the previous movement, testing the next number
    # ------------------------------------------

    def next_number(self, number):
     
        true_empty_cells = ept_cells
        keep_solving = True

        #while(ept_cells != []):
        # Traded this line to check the implicit boleaness bool(keep_solving)
        while(keep_solving):

            if ept_cells:
                next_move = ept_cells.pop(0)
                print "Next move: {}".format(next_move)
                
                self.board[next_move] = number
                print "Board in {}: {}\n".format(next_move, self.board[next_move])

                while check_solving() < 0:
                    self.board[next_move]

                #print self.peers[next_move]
            else:
                keep_solving = False

sudoku.py
#!/usr/bin/env python
# Four spaces as indentation [no tabs]

import sys
# Collections have the 'OrderedDict()' function
import collections
from common import *
from util import *
from ai_player import *

# ==========================================
# Sudoku
# ==========================================

numbers = [1,2,3,4,5,6,7,8,9
]
class Sudoku:

    # ------------------------------------------
    # Initialize
    # ------------------------------------------

    def __init__(self, human=True):
        self.human = human
        self.generate_board()
        self.reset()

    # ------------------------------------------
    # Reset
    # ------------------------------------------

    def reset(self):

        # If board receives a number in string type it will create an error in 
        # find_collision() function, but I don't know if there will be another error with 
        # this approach (int type).
        self.new_board = [0, 0, 8, 2, 9, 5, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 6, 0, 0, 9, 0, 0, 3, 0, 4, 0, 0, 6, 7, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 1, 0, 9, 0, 0, 8, 0, 0, 9, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 4, 3, 8, 9, 0, 0]

        for k in self.board:
            self.board[k] = self.new_board.pop(0)


    # ------------------------------------------
    # Generate rows, columns and blocks
    # ------------------------------------------

    def cross(self, A, B):

        return [a+b for a in A for b in B]


    def generate_board(self):

        digits = '123456789'
        cols = digits
        rows = 'ABCDEFGHI'

        # List of all cells:
        self.squares = self.cross(rows, cols)

        # List of units (columns, rows and blocks):
        self.unitlist = ([self.cross(rows, c) for c in cols] +
                    [self.cross(r, cols) for r in rows] +
                    [self.cross(rs, cs) for rs in ('ABC','DEF','GHI') for cs in ('123','456','789')])

        # -Dictionary with units
        # Conjunts where the cell is part of, counting itself.
        # Example: I6 = [[I1 - I9], [A6 - I6], [G4 - G6, H4 - H6, I4 - I6]]
        self.units = collections.OrderedDict((s, [u for u in self.unitlist if s in u]) for s in self.squares)

        # -Dictionary with peers:
        # List of every cell that is assigned in a unit with this cell, except itself.
        # Example: I6 = [A6, B6, C6, D6, E6, F6, G6, H6, I1, I2, I3, I4, I5, I7, I8, I9, G4, G5, H4, H5]
        self.peers = collections.OrderedDict((s, set(sum(self.units[s],[]))-set([s])) for s in self.squares)

        # Board with numbers played by player or given from the start
        self.board = collections.OrderedDict((s, None) for s in self.squares)


    # ------------------------------------------
    # Update
    # ------------------------------------------

    def update(self):

        # Board is an attribute, keep it safe sending a duplicate
        actual_board = self.board


        # If player is human:
        if self.human:
            print_board(actual_board)
            cell = str(raw_input("Cell name is made by letter from A to I and digit 1 to 9\n>> ")).upper()
            number = int(raw_input("Number between 1 and 9\n>> "))


            # Apply movement if there is one
            if (number != None) and (cell != None):

                # Validate number and cell
                if (1 <= number <= 9) and (cell in self.squares):

                    # Verifies if cell is clear
                    # IF retirado, para que seja possivel alterar o numero em qualquer momento
                    #if actual_board[cell] == 0:
                        actual_board[cell] = number

                        # Clear the CMD window
                        os.system('cls')

                        # Print actual board
                        print "\nPlayer chose to put number {} in cell {}.".format(number, cell)
                        #if self.debug:
                            #print_board(actual_board)

                        # Find collision
                        if find_collision(actual_board, self.peers, cell, number):
                            actual_board[cell] = 0
                            print "\nCollision found in cell {} by number played: {}.".format(cell, number)

                    # ELSE retirado, para que seja possivel alterar o numero em qualquer momento
                    #else:
                        #raise BlockedMovementException(actual_board, cell)
                else:
                    raise InvalidMovementException(actual_board, cell)
            else:
                raise NoMovementException(actual_board)


        # If player is not human, then it's the AI:
        else:
            
            ept_cells = find_empty_cells(self.board, self.squares)

            my_ai = AI_Player(actual_board, self.unitlist, self.peers, self.squares)
            my_ai.solve(ept_cells)
            
            print_board(actual_board)

            return 1

        return -1

# ==========================================
# Main
# ==========================================

if __name__ == "__main__":

    import argparse

    parser = argparse.ArgumentParser(description="Execute sudoku.")
    parser.add_argument("--human", action="store_true", help="AI player")

    args = parser.parse_args()

    # Create game with chosen players
    game = Sudoku(human=not args.human)

    # Clear the CMD window
    os.system('cls')

    # Print welcome
    print "\nWelcome to AI-SUDOKU\n"

    while game.update() < 0:
        print "THE GAME IS NOT FINISHED"

    print "THE GAME IS FINISHED!!!"